python/lsst/ts/tcpip/base_client_or_server.py

```diff
@@ -0,0 +1,478 @@
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+__all__ = ["BaseClientOrServer", "ConnectCallbackType"]
+
+import abc
+import asyncio
+import collections.abc
+import ctypes
+import inspect
+import logging
+import types
+import typing
+import warnings
+
+from . import utils
+from .constants import DEFAULT_MONITOR_CONNECTION_INTERVAL
+
+
+class BaseClientOrServer(abc.ABC):
+    """Abstract base class for a TCP/IP client or server.
+
+    Manage a stream reader and writer:
+
+    * Provide high-level methods for reading and writing.
+    * Optionally call a callback function when the connection state changes.
+    * Optionally monitor the connection in the background.
+
+    Parameters
+    ----------
+    log : `logging.Logger`
+        Logger.
+    connect_callback : callable or `None`, optional
+        Asynchronous or (deprecated) synchronous function to call when
+        the connection state changes.
+        If the other end closes the connection, it may take
+        ``monitor_connection_interval`` seconds or longer to notice.
+        The function receives one argument: this `BaseClientOrServer`.
+    monitor_connection_interval : `float`, optional
+        Interval between checking if the connection is still alive (seconds).
+        Defaults to DEFAULT_MONITOR_CONNECTION_INTERVAL.
+        If ≤ 0 then do not monitor the connection at all.
+        Monitoring is only useful if you do not regularly read from the reader
+        using the read methods of this class (or copying what they do
+        to detect and report hangups).
+    name : `str`
+        Optional name used for log messages.
+    **kwargs : `dict` [`str`, `typing.Any`]
+        Keyword arguments for start_task.
+
+    Attributes
+    ----------
+    log : `logging.Logger`
+        A child of the ``log`` constructor argument.
+    name : `str`
+        The ``name`` constructor argument.
+    reader : `asyncio.StreamReader` or None
+        Stream reader to read data from the server.
+        This will be a stream reader (not None) if `connected` is True.
+    writer : `asyncio.StreamWriter` or None
+        Stream writer to write data to the server.
+        This will be a stream writer (not None) if `connected` is True.
+    start_task : `asyncio.Future`
+        Future that is set done when the connection is made.
+    done_task : `asyncio.Future`
+        Future that is set done when this client is closed, at which point
+        it is no longer usable.
+    should_be_connected : `bool`
+        True if the connection was made and close not called.
+        The connection was unexpectedly lost if ``should_be_connected``
+        is true and ``connected`` is false (unless you close the connection
+        by calling `basic_close` or manually closing ``writer``).
```

Comment:
If I were a user of this class, I would be very confused between `should_be_connected` and `connected`. Instead, I would prefer `connected` that gives me a "clear" answer of connection status. I think this is what a user will expect.

Reply:
Good point. I have revised the documentation in order to try to clarify this.